stan/math/prim/mat/functor/map_rect_concurrent.hpp

```diff
@@ -0,0 +1,118 @@
+#ifndef STAN_MATH_PRIM_MAT_FUNCTOR_MAP_RECT_CONCURRENT_HPP
+#define STAN_MATH_PRIM_MAT_FUNCTOR_MAP_RECT_CONCURRENT_HPP
+
+#include <stan/math/prim/mat/fun/typedefs.hpp>
+
+#include <stan/math/prim/mat/functor/map_rect_reduce.hpp>
+#include <stan/math/prim/mat/functor/map_rect_combine.hpp>
+
+#include <vector>
+#include <thread>
+#include <future>
+#include <cstdlib>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+template <int call_id, typename F, typename T_shared_param,
+          typename T_job_param>
+Eigen::Matrix<typename stan::return_type<T_shared_param, T_job_param>::type,
+              Eigen::Dynamic, 1>
+map_rect_concurrent(
+    const Eigen::Matrix<T_shared_param, Eigen::Dynamic, 1>& shared_params,
+    const std::vector<Eigen::Matrix<T_job_param, Eigen::Dynamic, 1>>&
+        job_params,
+    const std::vector<std::vector<double>>& x_r,
+    const std::vector<std::vector<int>>& x_i, std::ostream* msgs = 0) {
+  typedef map_rect_reduce<F, T_shared_param, T_job_param> ReduceF;
+  typedef map_rect_combine<F, T_shared_param, T_job_param> CombineF;
+
+  const int num_jobs = job_params.size();
+  const vector_d shared_params_dbl = value_of(shared_params);
+  std::vector<std::future<std::vector<matrix_d>>> futures;
+
+  auto chunk_job = [&](int start, int end) -> std::vector<matrix_d> {
+    const int size = end - start;
+    std::vector<matrix_d> chunk_f_out;
+    chunk_f_out.reserve(size);
+    for (int i = start; i != end; i++)
+      chunk_f_out.push_back(ReduceF()(
+          shared_params_dbl, value_of(job_params[i]), x_r[i], x_i[i], msgs));
+    return chunk_f_out;
+  };
+
+  int num_threads = 1;
+
+#ifdef STAN_THREADS
+  const char* env_stan_num_threads = std::getenv("STAN_NUM_THREADS");
+  if (env_stan_num_threads != nullptr) {
+    const int env_num_threads = std::atoi(env_stan_num_threads);
+    if (env_num_threads > 0)
+      num_threads = env_num_threads;
+    else if (env_num_threads == -1)
+      num_threads = std::thread::hardware_concurrency();
+    // anything else will use 1 thread.
+  }
+#endif
+
+  const int num_jobs_per_thread = num_jobs / num_threads;
+  int num_jobs_per_thread_remainder = num_jobs % num_threads;
+  const int num_jobs_min = num_threads - num_jobs_per_thread_remainder;
+  int job_start = 0;
+  for (int j = 0; j < num_threads; j++) {
+    int job_end = job_start + num_jobs_per_thread;
+    // the excess jobs are assigned to the last processes
+    if (j >= num_jobs_min && num_jobs_per_thread_remainder > 0) {
+      job_end++;
+      num_jobs_per_thread_remainder--;
+    }
+    if (j == num_threads - 1)
+      job_end = num_jobs;
+    // we only defer the first chunk such that the main thread is not
+    // blocking if we use threading
+    futures.emplace_back(std::async(j == 0 ? std::launch::deferred :
+#ifndef STAN_THREADS
```

Comment:
Your way of coding this up is cleaner, so let's go with that. I would prefer to call the `n_threads` function `get_num_threads`; does that make sense to you? This would be more consistent with the `num` prefix and discriminate the function from the variable.

Also... I forgot to use the `nullptr` for the msgs argument of the function.

Reply:
One more note: The way you have setup things will have the main process do num_jobs_per_thread and the second thread must calculate num_jobs_per_thread+1 tasks. However, since we collect things in order the second thread will block until done. Thus, I think it is better to move the excess jobs to the end of queue. This is probably minor in most cases, but could yield better performance, I think.

Reply:
Ok, that makes sense.   I think that's a one line change.